WIP:Operation:deploy:Operations to deploy docker container on webhook (#570)
intel/dffml: docs/usage/webhook/deploy.rst

========================

.. note::

    All the code for this example is located under the
    `examples/ffmpeg <https://github.com/intel/dffml/blob/master/examples/ffmpeg/>`_
    directory of the DFFML source code.Rest of the tutorial is ran from the root of
    this folder.

Write operations and definitions to convert videos files to gif by calling
``ffmpeg`` (Make sure you `download and install <https://www.ffmpeg.org/download.html>`_ it.).
The operation accepts bytes (of the video), converts it to gif and outputs it as bytes.

**ffmpeg/operations.py**

.. literalinclude:: /../examples/ffmpeg/ffmpeg/operations.py

Dataflow and Config files
-------------------------

**ffmpeg/dataflow.py**

.. literalinclude:: /../examples/ffmpeg/ffmpeg/dataflow.py

.. code-block:: console

    $ mkdir -p deploy/mc/http deploy/df
    $ dffml service dev export -config yaml ffmpeg.dataflow:DATAFLOW > deploy/df/ffmpeg.yaml

Create the config file for the http server
in ``deploy/mc/http/ffmpeg.yaml``

.. code-block:: console

    $ cat > ./deploy/mc/http/ffmpeg.yaml <<EOF
    path: /ffmpeg
    input_mode: bytes:Input_file
    output_mode: bytes:post_input.Output_file
    EOF

- ``input_mode``

>     - ``bytes:Input_file``

Reviewer: ```suggestion

    - ``bytes:Input_file``
```

>         - we want the input from the request to be treated as bytes with definition ``Input_file``.

Reviewer: Capitalization

```suggestion
        - We want the input from the request to be treated as bytes with definition ``Input_file``.
```

- ``output_mode``
    - ``bytes:post_input.Output_file``
        - we want the response to be bytes taken from  ``results["post_input"]["Output_file"]``, where ``results`` is the output of the dataflow.

For more details see `HttpChannelConfig <../../plugins/service/http/dataflow.html#HttpChannelConfig>`__ .

.. _usage_ffmpeg_deploy_serve:

Serving the DataFlow
--------------------

Serving the dataflow on port 8080

.. code-block:: console

    $ dffml service http server -insecure -mc-config deploy -port 8080

.. warning::

    The ``-insecure`` flag is only being used here to speed up this
    tutorial. See documentation on HTTP API
    :doc:`/plugins/service/http/security` for more information.

> 

Reviewer: We should probably tell people to `curl` down the video that you've added to the GitHub repo. Or use their own if they have one.

Now from another terminal, we can send post requests to the dataflow running at this port.

.. code-block:: console

>     $ curl -v --request POST --data-binary @input.mp4 http://localhost:8080/ffmpeg -o output.gif

Reviewer: DOOOOOOPE

This command will convert input.mp4 to output.gif

Deploying on docker container
=============================

A ``Dockerfile`` is already generated in ffmpeg folder. We need to modify it to include ``ffmpeg``.

**Dockerfile**

.. literalinclude:: /../examples/ffmpeg/Dockerfile

.. note::
    The run command in the comment section of the Dockerfile will be used to execute
    the container after receving webhooks , so make sure you change it to your usecase.

For this tutorial we will change it to

.. code-block:: json

>     docker run --rm -d -ti -p 8080:8080 aghinsa/ffmpeg -mc-config deploy -insecure -log debug

Reviewer: The container will have been built with their username

```suggestion
    docker run --rm -d -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug
```

.. note::

    The image built after pulling the contaier will be taged ``USERNAME/REPONAME``, where USERNAME and REPONAME
    are gathered from the github html url, received in the webhook.

We can run the container and sent a post request to verify that the container is working.

.. code-block:: console

    $ docker build -t aghinsa/ffmpeg .

>     $ docker run --rm -d -ti -p 8080:8080 aghinsa/ffmpeg -mc-config deploy -insecure -log debug

Reviewer: Same thing, `$USER` should evaluate to the correct username 

```suggestion
    $ docker run --rm -d -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug
```

.. code-block:: console

    $ curl -v --request POST --data-binary @input.mp4 http://localhost:8080/ffmpeg -o output.gif

> Now in :ref:`usage_webhook` we'll setup this container to be redployed .

Reviewer: ```suggestion
Now in :ref:`usage_webhook` we'll setup this container to be automatically redeployed
whenever we push to the Git repo containing this code.
```

.. _usage_ffmpeg_deploy:

> In this tutorial we will deploy a dataflow(ffmpeg dataflow) which converts a video to gif over an http server.We'll also

Reviewer: ```suggestion
In this tutorial we will deploy a dataflow (ffmpeg dataflow) which converts a video to gif over an http server. We'll also
```

> see how to deploy the same http server in a docker container.Finally in :ref:`usage_ffmpeg_deploy_serve`

Reviewer: ```suggestion
see how to deploy the same http server in a docker container. Finally in :ref:`usage_ffmpeg_deploy_serve`
```

> we'll setup another http server which waits on gtihub webhooks to rebuilt and deploy the ffmpeg dataflow.

Reviewer: ```suggestion
we'll setup another http server which waits on GitHub webhooks to rebuilt and deploy the ffmpeg dataflow.
```

> Deploying on http server

Reviewer: ```suggestion
Deploying with the HTTP Service
```

Reviewer: The description above this should go directly below this

> ========================

Reviewer: ```suggestion
===============================
```

.. note::

    All the code for this example is located under the
    `examples/ffmpeg <https://github.com/intel/dffml/blob/master/examples/ffmpeg/>`_
    directory of the DFFML source code.Rest of the tutorial is ran from the root of
    this folder.

Write operations and definitions to convert videos files to gif by calling
``ffmpeg`` (Make sure you `download and install <https://www.ffmpeg.org/download.html>`_ it.).

> The operation accepts bytes (of the video), converts it to gif and outputs it as bytes.

Reviewer: We need to be sure to tell them to create the package

```suggestion
The operation accepts bytes (of the video), converts it to gif and outputs it as bytes.

Create the Package
------------------

To create a new source we first create a new Python package. DFFML has a script
to do it for you.

.. code-block:: console

    $ dffml service dev create operations ffmpeg
    $ cd ffmpeg

We will start writing our source in ``./ffmpeg/operations.py``
```

**ffmpeg/operations.py**

.. literalinclude:: /../examples/ffmpeg/ffmpeg/operations.py

Dataflow and Config files
-------------------------

> **ffmpeg/dataflow.py**

Reviewer: I don't see this file? Also, shouldn't this really be the part where we add the operation to `ffmpeg/setup.py` `entry_points`, re-install the package, and then use the `dffml dataflow create` command? Instead of `export`.

Reviewer: Probably don't need the `dataflow.py` file anyway is what I'm saying

Author: sorry i forgot to commit that file.I added `resolution` as a seed input,so that we can use `input_mode:bytes` and just post the file.
which one would be better, current or sending `resolution` and `input_file` with `input_mode:default`?

Reviewer: Keep resolution in the `seed`. We'll have to implement multipart form encoding to do that correctly with `default`

.. literalinclude:: /../examples/ffmpeg/ffmpeg/dataflow.py

.. code-block:: console

    $ mkdir -p deploy/mc/http deploy/df
    $ dffml service dev export -config yaml ffmpeg.dataflow:DATAFLOW > deploy/df/ffmpeg.yaml

> Create the config file for the http server

Reviewer: ```suggestion
Create the config file for the HTTP service
```

in ``deploy/mc/http/ffmpeg.yaml``

.. code-block:: console

    $ cat > ./deploy/mc/http/ffmpeg.yaml <<EOF
    path: /ffmpeg

>     input_mode: bytes:Input_file

Reviewer: We should probably name `Input_file` -> `input_file`, since we usually don't have mixed case definition `name`s

```suggestion
    input_mode: bytes:input_file
```

>     output_mode: bytes:post_input.Output_file

Reviewer: Same with output file

```suggestion
    output_mode: bytes:post_input.output_file
```

    EOF

- ``input_mode``

>     - ``bytes:Input_file``

Reviewer: ```suggestion
    - ``bytes:input_file``
```

>         - We want the input from the request to be treated as bytes with definition ``Input_file``.

Reviewer: ```suggestion
        - We want the input from the request to be treated as bytes with definition ``input_file``.
```

- ``output_mode``

>     - ``bytes:post_input.Output_file``

Reviewer: ```suggestion
    - ``bytes:post_input.output_file``
```

>         - We want the response to be bytes taken from  ``results["post_input"]["Output_file"]``, where ``results`` is the output of the dataflow.

Reviewer: ```suggestion
        - We want the response to be bytes taken from  ``results["post_input"]["output_file"]``, where ``results`` is the output of the dataflow.
```

.. _usage_ffmpeg_deploy_serve:

Serving the DataFlow
--------------------

Serving the dataflow on port 8080

.. code-block:: console

    $ dffml service http server -insecure -mc-config deploy -port 8080

.. warning::

    The ``-insecure`` flag is only being used here to speed up this
    tutorial. See documentation on HTTP API
    :doc:`/plugins/service/http/security` for more information.

Now from another terminal, we can send post requests to the dataflow running at this port.

.. code-block:: console

    $ curl -v --request POST --data-binary @input.mp4 http://localhost:8080/ffmpeg -o output.gif

You should replace ``input.mp4`` with path to your video file and ``output.gif`` to where you want the converted gif
to be output to. An example video is available `here <https://github.com/intel/dffml/blob/master/examples/ffmpeg/>`_ .

Deploying on docker container
=============================

A ``Dockerfile`` is already generated in ffmpeg folder. We need to modify it to include ``ffmpeg``.

**Dockerfile**

.. literalinclude:: /../examples/ffmpeg/Dockerfile

.. note::
    The run command in the comment section of the Dockerfile will be used to execute

>     the container after receving webhooks , so make sure you change it to your usecase.

Reviewer: ```suggestion
    the container after receving webhooks, so make sure you change it to your usecase.
```

For this tutorial we will change it to

> .. code-block:: json

Reviewer: ```suggestion
.. code-block:: Dockerfile
```

>     docker run --rm -d -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug

Reviewer: I believe there's a comment on this line right

```suggestion
    # docker run --rm -d -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug
```

.. _usage_ffmpeg_deploy:

Deploying with the HTTP Service
===============================

In this tutorial we will deploy a dataflow(ffmpeg dataflow) which converts a video to gif over an HTTP service. We'll
also see how to deploy the same in a docker container. Finally in :ref:`usage_ffmpeg_deploy_serve`
we'll setup another HTTP service which waits on GitHub webhooks to rebuilt and deploy the ffmpeg dataflow.

.. note::

    All the code for this example is located under the
    `examples/ffmpeg <https://github.com/intel/dffml/blob/master/examples/ffmpeg/>`_

>     directory of the DFFML source code.Rest of the tutorial is ran from the root of

Reviewer: ```suggestion
    directory of the DFFML source code.
```

>     this folder.

Reviewer: ```suggestion
```

> 

Reviewer: Let's have them install `dffml-service-http` and `dffml-config-yaml` after the `.. note`

``ffmpeg`` (Make sure you `download and install <https://www.ffmpeg.org/download.html>`_ it.).
The operation accepts bytes (of the video), converts it to gif and outputs it as bytes.
We will start writing our operation in ``./ffmpeg/operations.py``

**ffmpeg/operations.py**

.. literalinclude:: /../examples/ffmpeg/ffmpeg/operations.py

Add the operation to ``ffmpeg/setup.py``

.. code-block:: python

    common.KWARGS["entry_points"] = {
        "dffml.operation": [
            f"convert_to_gif = {common.IMPORT_NAME}.operations:convert_to_gif"
        ]
    }

Install the package

.. code-block:: console

    $ pip install -e .

Dataflow and Config files
-------------------------

**ffmpeg/dataflow.py**

.. literalinclude:: /../examples/ffmpeg/ffmpeg/dataflow.py

.. code-block:: console

    $ mkdir -p deploy/mc/http deploy/df
    $ dffml service dev export -config yaml ffmpeg.dataflow:DATAFLOW > deploy/df/ffmpeg.yaml

Create the config file for the HTTP service
in ``deploy/mc/http/ffmpeg.yaml``

.. code-block:: console

    $ cat > ./deploy/mc/http/ffmpeg.yaml <<EOF
    path: /ffmpeg
    input_mode: bytes:input_file

>     output_mode: bytes:application/octet-stream:post_input.output_file

Reviewer: ```suggestion
    output_mode: bytes:image/gif:post_input.output_file
```

    EOF

- ``input_mode``

    - ``bytes:input_file``

        - We want the input from the request to be treated as bytes with definition ``input_file``.

- ``output_mode``

>     - ``bytes:application/octet-stream:post_input.output_file``

Reviewer: ```suggestion
    - ``bytes:image/gif:post_input.output_file``
```

>         - We want the response (content_type = application/octet-stream) to be bytes taken from  ``results["post_input"]["output_file"]``,

Reviewer: ```suggestion
        - We want the response (content_type = image/gif) to be bytes taken from  ``results["post_input"]["output_file"]``,
```

          where ``results`` is the output of the dataflow.

For more details see `HttpChannelConfig <../../plugins/service/http/dataflow.html#HttpChannelConfig>`__ .

.. _usage_ffmpeg_deploy_serve:

Serving the DataFlow
--------------------

Serving the dataflow on port 8080

.. code-block:: console

    $ dffml service http server -insecure -mc-config deploy -port 8080

.. warning::

    The ``-insecure`` flag is only being used here to speed up this
    tutorial. See documentation on HTTP API
    :doc:`/plugins/service/http/security` for more information.

Now from another terminal, we can send post requests to the dataflow running at this port.

.. code-block:: console

    $ curl -v --request POST --data-binary @input.mp4 http://localhost:8080/ffmpeg -o output.gif

You should replace ``input.mp4`` with path to your video file and ``output.gif`` to where you want the converted gif

> to be output to. An example video is available `here <https://github.com/intel/dffml/blob/master/examples/ffmpeg/>`_ .

Reviewer: ```suggestion
to be output to. An example video is available `here <https://github.com/intel/dffml/raw/master/examples/ffmpeg/tests/input.mp4>`_ .
```

Reviewer: Below this line let's put a `.. warning` about how we've noticed that `snap install ffmpeg` did not work.

> Deploying on docker container

Reviewer: ```suggestion
Deploying via container
```

> =============================

Reviewer: ```suggestion
=======================
```

A ``Dockerfile`` is already generated in ffmpeg folder. We need to modify it to include ``ffmpeg``.

**Dockerfile**

.. literalinclude:: /../examples/ffmpeg/Dockerfile

> .. note::

Reviewer: ```suggestion
.. note::

```

    The run command in the comment section of the Dockerfile will be used to execute
    the container after receving webhooks, so make sure you change it to your usecase.

For this tutorial we will change it to

.. code-block:: Dockerfile

>     # docker run --rm -d -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug

Reviewer: ```suggestion
    # docker run --rm -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug
```

We probably want to have the operations which restart the container add in `-d` and `--restart=always`

.. note::

    The image built after pulling the contaier will be taged ``USERNAME/REPONAME``, where USERNAME and REPONAME
    are gathered from the github html url, received in the webhook.

We can run the container and sent a post request to verify that the container is working.

.. code-block:: console

    $ docker build -t $USER/ffmpeg .

>     $ docker run --rm -d -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug

Reviewer: ```suggestion
    $ docker run --rm -ti -p 8080:8080 $USER/ffmpeg -mc-config deploy -insecure -log debug
```

> 

Reviewer: ```suggestion

Now, in another terminal

```

.. _usage_ffmpeg_deploy:

Deploying on http server
========================

> 

Reviewer: Let's make the first section a quick show of the operation, and then move into the dataflow stuff.

```suggestion

.. note::

    All the code for this example is located under the
    `examples/ffmpeg <https://github.com/intel/dffml/blob/master/examples/ffmpeg/>`_
    directory of the DFFML source code.

Writing operations and definitions to convert videos files to gif by calling ``ffmpeg``

**ffmpeg/operations.py**

.. literalinclude:: /../examples/ffmpeg/ffmpeg/operations.py

```

Dataflow and Config files
-------------------------

Create a ``Dataflow`` with operation ``convert_to_gif`` and save it in ``deploy/df/ffmpeg.yaml``

.. code-block:: console

    $ mkdir -p deploy/mc/http deploy/df
    $ cat > /tmp/operations <<EOF
    convert_to_gif
    EOF

    $ dffml dataflow create -config yaml $(cat /tmp/operations) > deploy/df/ffmpeg.yaml

Create the `Config <../../plugins/service/http/dataflow.html#HttpChannelConfig>`__ file for the http server
in ``deploy/mc/http/ffmpeg.yaml``

.. code-block:: console

    $ cat > ./deploy/mc/http/ffmpeg.yaml <<EOF
    path: /ffmpeg

>     presentation: json

Reviewer: Remember that we want the output to be the bytes of the `output.gif` file written directly to the client (preferably with a `Content-Type` header set to whatever is specified in `ffmpeg.yaml` as `content_type`). And we want to take the post body and treat it as the input file. Ideally we'd have it as `stream` and pass that stream as the stdin to the call to subprocess. However, we may have write those bytes out to a temporary file, and then pass the temporary file name into the ffmpeg operation.

Author: @pdxjohnny by `content_type` do you mean,`presentation`,or should i add another param?

Author: also when you say we write the output file back to the client,we'd have to have an option to do
```
return web.Response(body = result,content_type = config.content_type)
```
in place of 
https://github.com/aghinsa/dffml/blob/bbfec63078863247b2bba4ff6ce60e7403cce4d7/service/http/dffml_service_http/routes.py#L267-L281

is that the idea?
if we are doing that,ig it makes sense to do [issue#608](https://github.com/intel/dffml/issues/608) along with this.

Reviewer: Yes, and yes we should do #608 along with this.

    asynchronous: false
    EOF

.. _usage_ffmpeg_deploy_serve:

Serving the DataFlow
--------------------

Serving the dataflow on port 8080

.. code-block:: console

    $ dffml service http server -insecure -mc-config deploy -port 8080

.. warning::

    The ``-insecure`` flag is only being used here to speed up this
    tutorial. See documentation on HTTP API
    :doc:`/plugins/service/http/security` for more information.

Now from another terminal, we can send post requests to the dataflow running at this port.

.. code-block:: console

    $ curl -s \

>   --header "Content-Type: application/json" \

Reviewer: ```suggestion
      --header "Content-Type: application/json" \
```

>   --request POST \

Reviewer: ```suggestion
      --request POST \
```

>   --data '{"convert": [{"value":"PATH_TO_INPUT","definition":"input_file"},

Reviewer: ```suggestion
      --data '{"convert": [{"value":"PATH_TO_INPUT","definition":"input_file"},
```

.. _usage_ffmpeg_deploy:

Deploying on http server

> ========================

Reviewer: We need a short description of what this tutorial is and why one might use it to go after the `========================`

Dataflow and Config files
-------------------------

Create a ``Dataflow`` with operation ``convert_to_gif`` and save it in ``deploy/df/ffmpeg.yaml``

.. code-block:: console

    $ mkdir -p deploy/mc/http deploy/df
    $ cat > /tmp/operations <<EOF
    convert_to_gif
    EOF

    $ dffml dataflow create -config yaml $(cat /tmp/operations) > deploy/df/ffmpeg.yaml

Create the `Config <../../plugins/service/http/dataflow.html#HttpChannelConfig>`__ file for the http server
in ``deploy/mc/http/ffmpeg.yaml``

.. code-block:: console

    $ cat > ./deploy/mc/http/ffmpeg.yaml <<EOF
    path: /ffmpeg
    presentation: json
    asynchronous: false
    EOF

.. _usage_ffmpeg_deploy_serve:

Serving the DataFlow
--------------------

Serving the dataflow on port 8080

.. code-block:: console

    $ dffml service http server -insecure -mc-config deploy -port 8080

.. warning::

    The ``-insecure`` flag is only being used here to speed up this
    tutorial. See documentation on HTTP API
    :doc:`/plugins/service/http/security` for more information.

> Now from another terminal,we can send post requests to the dataflow running at this port.

Reviewer: ```suggestion
Now from another terminal, we can send post requests to the dataflow running at this port.
```
